0000-macro-value.md

```diff
@@ -0,0 +1,300 @@
+    };
+}
+```
+
+This time,
+`$a` and `$b` are evaluated a single time upon invoking the macro,
+and each expansion of the capture refers to the same value,
+just like function arguments.
+
+(no need to mention temporary lifetimes in the guide.)
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A new macro matching mode,
+`$:value`,
+is added.
+It captures the same grammar,
+has the same follow set,
+and can be expanded in the same positions as `$:expr`.
+
+A `macro_rules` macro capturing an expression as `$:value` can only be used in expression position,
+not any other position
+(item, type, etc.).
+As such,
+extra information is provided to the compiler that it MAY use for nicer error messages.
+(When expanding an expression-position-only macro in item position,
+ the current 1.51 rustc says
+ "the usage of `mac!` is _likely_ invalid in item context"
+ (emphasis mine),
+ which could be strengthened if all macro arms capture `$:value`.)
+
+For a given capture `$name:value`,
+the captured expression is evaluated a single time upon entry into the macro expansion,
+whether `$name` is mentioned in the macro expansion zero, one, or any number times.
+Every expansion of `$name` within the macro expansion refers to the name of the temporary where the captured expression was evaluated.
+If more than one `$:value` capture is present,
+they are evaluated from left to right.
+The intent is that this has identical semantics to that of a function argument capture.
+
+A compiler MAY implement this by expanding to a `match` expression,
+similar to the following:
+
+```rust
+macro_rules! mac! {
+    /* other arms */
+    ( /* other captures */ $name:value /* other captures */ ) => {
+        /* macro body */
+    };
+    /* other arms */
+}
+// "desugars" to
+macro_rules! mac! {
+    /* other arms */
+    ( /* other captures */ $name:expr /* other captures */ ) => {
+        match $name {
+            name => {
+                /* macro body, $name replaced with name (hygienically) */
+            }
+        }
+    };
+    /* other arms */
+}
+```
+
+but the compiler is expected to also handle the case where `$:value` is inside of a macro repetition,
+which cannot be directly implemented by just a desugaring of the `macro_rules!` invocation.
+
+To illustrate (one of) the differences in lifetimes, consider temporary lifetime extension:
+
+```rust
+macro_rules! mlet {
+    ( $e:expr ) => {{
+        let e = $e;
+        e
+    }};
+}
+
+macro_rules! mmatch {
+    ( $e:expr ) => {
+        match $e {
+            e => e,
+        }
+    };
+}
+
+fn mfn<T>(e: T) -> T {
+    e
+}
+
+struct Wrap(&'static str);
+impl Drop for Wrap {
+    fn drop(&mut self) {}
+}
+
+fn a() { let _a = mlet!  (&Wrap("a")); } // borrowck error
+fn b() { let _b = mmatch!(&Wrap("b")); } // compiles
+fn c() { let _c = mfn    (&Wrap("c")); } // compiles
+```
+
+plus there are other differences with drop timing of temporaries,
```

Comment:
It would be best to explain those differences.

Reply:
I know and.... I'm not perfectly aware of what they all are, thus the sidestep here.